// novelapi/src/Generation.js
import fetch from 'node-fetch';

export class Generator
{
    auth;
    defaultModel = 'euterpe-v2'
    constructor(auth)
    {
        this.auth = auth;
    }

    /**
     * What this module is built for. AI text generation.
     * @param {String} input Input text for generation. Maximum length is 14000 characters.
     * @param {String} model Model for text generation. Available models are: [ 2.7B, 6B-v4, euterpe-v2, genji-python-6b, genji-jp-6b, genji-jp-6b-v2, krake-v1 ]
     * @param {String} customModel Model for text generation. Available models are: [ 2.7B, 6B-v4, euterpe-v2, genji-python-6b, genji-jp-6b, genji-jp-6b-v2, krake-v1 ]
     * @param {*} customParamaters Generation parameters. Defaults are based on Genisis preset, but can be adjusted or fed through here. Default presets based on Genisis, additional options can be found in the parameters section of AiGenerateRequest: https://api.novelai.net/docs/static/index.html#/
     * @returns JSON: Generated Text
     */
    async Generate(input, customModel, customParamaters)
    {
        return await GGenerateText(this.auth, input, customModel, customParamaters)
    }
    
    /**
     * Classifies text based on content. Returns score based on explict words used. Does not function on retail clients. Last checked 03/25/2022
     * @param {String} input Text to classify.
     * @param {String} customParamaters Text to classify.
     * @param {String} customModel Model to use.
     * 
     * @returns JSON
     */
    async Classify(input)
    {
        return await GClassify(this.auth, input, customModel);
    }
}

export async function GGenerateText(auth, input, model, parameters) {
    try {
        const response = await fetch('https://api.novelai.net/ai/generate', {
            method: 'POST',
            headers: { 'Authorization': 'Bearer ' + auth, 'Content-Type': 'application/json' },
            body: JSON.stringify({ 'input': input, 'model': model, 'parameters': parameters })
        });
        //console.log(response)
        return await response.json().catch(error => {console.error(`NovelAPI fetch catch: ${error}`);});
    } catch (error) {
        console.error(error);
        return null;
    }
}

export async function GClassify(auth, input) {
    try {
      const response = await fetch('https://api.novelai.net/ai/classify', {
        method: 'POST',
        headers: {
          'Authorization': 'Bearer ' + auth,
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({ 'input': input })
      }); 
      //console.log(response)
      return await response.json().catch(error => {console.error(`NovelAPI fetch catch: ${error}`);});
    } catch (error) {
      console.error(error);
      return null;
    }
  }
